mlir/lib/Quantum/Transforms/PropagateSimpleStatesAnalysis.hpp

```diff
@@ -0,0 +1,264 @@
+        // in |0>. Then the FSM propagates the states through the custom op gates
+
+        target->walk([&](quantum::ExtractOp op) {
+            // Do not analyze any operations in invalid ops or regions.
+            // The only valid ops are the custom/extract ops whose immediate parent is the qnode
+            // function. With this, we skip over regions like control flow.
+            if (!isImmediateChild(op, target)) {
+                return;
+            }
+
+            // starting qubits are in |0>
+            // We restrict "starting qubits" to those who are extracted immediately from alloc ops
+            if (isa<quantum::AllocOp>(op->getOperand(0).getDefiningOp())) {
+                qubitValues[op->getResult(0)] = QubitState::ZERO;
+                return;
+            }
+        });
+
+        target->walk([&](quantum::CustomOp op) {
+            if (!isImmediateChild(op, target)) {
+                return;
+            }
+
+            if (op->getNumResults() != 1) {
+                // restrict to single-qubit gates
+                return;
+            }
+
+            Value res = op->getResult(0);
+
+            // takes in parameters other than the parent qubit
+            // e.g. the rotation angle
+            // must be NOT_A_BASIS!
+            if (op->getNumOperands() != 1) {
+                qubitValues[res] = QubitState::NOT_A_BASIS;
+                return;
+            }
+
+            // get state from parent and gate
+            StringRef gate = cast<quantum::CustomOp>(op).getGateName();
+            Value parent = op->getOperand(0);
+
+            // Unknown parent state, child state is thus also unknown
+            if (!qubitValues.contains(parent) || isOther(qubitValues[parent])) {
+                qubitValues[res] = QubitState::NOT_A_BASIS;
+                return;
+            }
+
+            // Identity preserves parent state
+            if (gate == "Identity") {
+                qubitValues[res] = qubitValues[parent];
+                return;
+            }
+
+            // A valid FSM transition gate
+            // Special treatment for S+ gate from |L> and |R>
+            if ((isLeft(qubitValues[parent]) || isRight(qubitValues[parent])) && gate == "S") {
+                if (op->hasAttr("adjoint")) {
+                    switch (qubitValues[parent]) {
+                    case QubitState::LEFT:
+                        qubitValues[res] = QubitState::PLUS;
+                        break;
+                    case QubitState::RIGHT:
+                        qubitValues[res] = QubitState::MINUS;
+                        break;
+                    default:
+                        // this will never trigger as the switch is inside an if
+                        break;
+                    }
+                }
+                else {
+                    qubitValues[res] = QubitState::NOT_A_BASIS;
+                }
+                return;
+            }
```

Comment:
Could this whole branch be simplified by updating the `gate` value to `S+` if we have `S` and `adjoint`, and then still use the table like below?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1154/commits/1badd0f8024aea29392e257b807662041580e296

Good idea! Definitely much cleaner.